fendermint/vm/topdown/src/sync.rs

```diff
@@ -242,106 +239,217 @@ async fn reset_cache<T: ParentFinalityStateQuery + Send + Sync + 'static>(
     Ok(())
 }
 
-/// Obtains the last recorded height from provider cache or from last committed finality height.
-async fn last_recorded_height(
+/// A util struct that tracks the last recorded height
+enum LastRecordedBlock {
+    FilledBlock {
+        height: BlockHeight,
+        hash: BlockHash,
+    },
+    NullBlock(BlockHeight),
+    Empty,
+}
+
+impl LastRecordedBlock {
+    fn filled(height: BlockHeight, hash: BlockHash) -> Self {
+        Self::FilledBlock { height, hash }
+    }
+
+    fn null(height: BlockHeight) -> Self {
+        Self::NullBlock(height)
+    }
+
+    fn empty() -> Self {
+        Self::Empty
+    }
+}
+
+/// Getting the last recorded block height/hash
+async fn last_recorded_data(
     provider: &Arc<Toggle<CachedFinalityProvider<IPCProviderProxy>>>,
 ) -> anyhow::Result<Option<(BlockHeight, BlockHash)>> {
-    let result = atomically(|| {
-        let h = if let Some(h) = provider.latest_height_hash()? {
-            Some(h)
-        } else if let Some(f) = provider.last_committed_finality()? {
-            Some((f.height, f.block_hash))
-        } else {
-            None
-        };
-        Ok(h)
-    })
-    .await;
+    match atomically(|| last_recorded_block(provider)).await {
+        LastRecordedBlock::Empty => Ok(None),
+        LastRecordedBlock::FilledBlock { height, hash } => Ok(Some((height, hash))),
+        LastRecordedBlock::NullBlock(height) => {
+            // Imaging the list of blocks as follows:
+            //
+            // F0  B0  B1  N0  N1  B2  B3  B4
+            //
+            // where F0 represents the last committed finality, B* represents non-null blocks and
+            // N* represents null blocks.
+            //
+            // Currently the last recorded block is N1, so the next block to sync in parent is B2.
+            // The response from getting block hash at height B2 from fvm eth apis would return:
+            //
+            // Block height: B2, Block hash: hash(B2), Parent block hash: hash(B1)
+            //
+            // F0  B0  B1  N0  N1  B2
+            //     B0' N0' N1' N2' B2  B3  B4       <====== reorged chain case 1
+            //     B0  B1  B2' B3' B2               <====== reorged chain case 2
+            //     B0  B1  N0' B1' B2               <====== reorged chain case 3
```

Comment:
It's not entire clear what's happening here. Are these observations on the syncer side, say, we have seen B1 followed by B2' then B3' then a switch back to B2 again?

Reply:
Okay so I got confused by the notation here. Above in `F0  B0  B1  N0  N1  B2  B3  B4` it looks like we use the `B*` numbers as identities for existing blocks, so the ID of the block at height 7 is 3. But then here we see `B3'` and `B1'` which I usually interpret as "a block at height 1 on a fork" to contrast it with `B1`. 

Perhaps it would help to draw down a chain with these forks and then mark them with identities as well. I think the above scenarios would translate to this:
```
F0 - B0 - B1 - N - N - B2 - B3 - B4
  |         |
  |         |- B8
  |         |
  |         |- B6 - B7              
  |
  |- B5 - N - N - N 
```

with the following translations, with `*` marking reorgs
```
F0  B0  B1  N0  N1  B2          => F0 B0 B1 N  N  B2
    B0' N0' N1' N2' B2  B3  B4  => F0 B5 N  N  N *B2 B3 B4
    B0  B1  B2' B3' B2          => F0 B0 B1*B6 B7*B2 
    B0  B1  N0' B1' B2          => F0 B0 B1 N  B8*B2      
```